### PR TITLE
[export] allow Dim(1,2) for export dynamic shapes (v2 after revert)

### PR DESCRIPTION
**Pull request by pianpwk**
Creating this after [PR](https://github.com/pytorch/pytorch/pull/121642) got reverted.

Current dynamic shapes implementation fixes lower range of Dims to be 2 for analysis, but allows 0/1 shapes during runtime. This leads to failures when initializing Dim(1,2). This PR sets the lower bound to 0, and avoids erroring out when conflicting with the generated (2, maxsize) constraint during analysis.

Also resolves a derived dim constraints issue with the following code:
```
class Bar(torch.nn.Module):
    def forward(self, x, y):
        return x + y[1:]

dx = Dim("dx", min=1, max=3)
ep = export(
    Bar(),
    (torch.randn(2, 2), torch.randn(3, 2)),
    dynamic_shapes=({0: dx, 1: None}, {0: dx+1, 1: None})
)
print(ep.range_constraints)
```

In main:
```
{s0: ValueRanges(lower=2, upper=3, is_bool=False), s0 + 1: ValueRanges(lower=3, upper=4, is_bool=False)}
```

This PR:
```
{s0: ValueRanges(lower=1, upper=3, is_bool=False), s0 + 1: ValueRanges(lower=2, upper=4, is_bool=False)}
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_export/non_strict_utils.py [hunk 1]
```diff
@@ -193,6 +198,18 @@ def make_constraints(
...
     #   - _trace.py installs shape metadata in IR.
 
+    if dynamic_shapes is None:
+        return {}
+
+    flat_dynamic_shapes, _ = tree_flatten(
+        dynamic_shapes,
+        is_leaf=lambda x: x is None
+        or all(
+            not isinstance(y, Iterable)
```
Comment (avikchaudhuri) [inline torch/_export/non_strict_utils.py:208]: you mean `isinstance(y, _Dim)` instead, no?

Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py:208]: Basically IIUC you're trying to keep things like `None` and `{0: dim}` as leaves.

**Review by avikchaudhuri — APPROVED** [review]
see nit

File: torch/_export/non_strict_utils.py (hunk 1 above)
Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py:208]: Actually look at `update_symbols` in `process_dynamic_shapes` for the patterns here. Another one is `(dim1, dim2)` or `[dim1, dim2]`.

Comment (pianpwk) [reply torch/_export/non_strict_utils.py:208]: @avikchaudhuri Hmm I think the best I can do is 
```
is_leaf=lambda x: x is None
    or all(
        isinstance(y, _Dim) or y is None
        for y in (x.values() if isinstance(x, dict) else x)
    )
```

Because I think we still want to support things like `({0: dim, 1: None}, None)`

**Review by zhxchen17 — APPROVED** [review]

File: torch/_export/non_strict_utils.py (hunk 1 above)
Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py]: Well tuples where everything is dim or None is also in scope for leaves, right?
So either None, or a dict where all values are dim or None, or a tuple where everything is dim or None?

Correct me if I'm wrong. Basically {0: dim, 1: None} can be alternatively written as (dim, None) I think. None can mean static dim or None can mean, this tensor is entirely made of static dims.

Comment (avikchaudhuri) [reply torch/_export/non_strict_utils.py]: Btw if the lambda is getting too complicated lol pull it out as a function.
